src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Browser.cs

```diff
@@ -64,43 +70,82 @@ private string JSGetNativeDisplayName(string localeName, string cultureName)
         {
-            int result = Interop.JsGlobalization.GetFirstDayOfWeek(localeName, out int exception, out object ex_result);
-            if (exception != 0)
+            ReadOnlySpan<char> localeNameSpan = localeName.AsSpan();
+            fixed (char* pLocaleName = &MemoryMarshal.GetReference(localeNameSpan))
             {
-                // Failed, just use 0
-                Debug.Fail($"[CultureData.GetFirstDayOfWeek()] failed with {ex_result}");
-                return 0;
+                nint exceptionPtr = Interop.JsGlobalization.GetFirstDayOfWeek(pLocaleName, localeNameSpan.Length, out int result);
+                if (exceptionPtr != IntPtr.Zero)
+                {
+                    int success = Helper.MarshalAndThrowIfException(
+                        exceptionPtr,
+                        failOnlyDebug: true,
+                        failureMessage: $"[CultureData.GetFirstDayOfWeek()] failed with");
+                    // Failed, just use 0
+                    if (success == -1)
+                        return 0;
+                }
+                return result;
             }
-            return result;
         }
 
         private static unsafe int GetFirstWeekOfYear(string localeName)
         {
-            int result = Interop.JsGlobalization.GetFirstWeekOfYear(localeName, out int exception, out object ex_result);
-            if (exception != 0)
+            ReadOnlySpan<char> localeNameSpan = localeName.AsSpan();
+            fixed (char* pLocaleName = &MemoryMarshal.GetReference(localeNameSpan))
+            {
+                nint exceptionPtr = Interop.JsGlobalization.GetFirstWeekOfYear(pLocaleName, localeNameSpan.Length, out int result);
+                if (exceptionPtr != IntPtr.Zero)
+                {
+                    int success = Helper.MarshalAndThrowIfException(
+                        exceptionPtr,
+                        failOnlyDebug: true,
+                        failureMessage: $"[CultureData.GetFirstWeekOfYear()] failed with");
+                    // Failed, just use 0
+                    if (success == -1)
+                        return 0;
+                }
+                return result;
+            }
+        }
+    }
+
+    internal static class Helper
+    {
+        internal static int MarshalAndThrowIfException(nint exceptionPtr, bool failOnlyDebug = false, string failureMessage = "")
+        {
+            if (exceptionPtr != IntPtr.Zero)
             {
-                // Failed, just use 0
-                Debug.Fail($"[CultureData.GetFirstDayOfWeek()] failed with {ex_result}");
-                return 0;
+                string message = Marshal.PtrToStringUni(exceptionPtr)!;
+                Marshal.FreeHGlobal(exceptionPtr);
+                if (failOnlyDebug)
+                {
+                    Debug.Fail($"{failureMessage} {message}");
+                    return -1;
+                }
+                throw new Exception(message);
```

Comment:
Should there be more specific exception type ? What is the type of exception when ICU fails ?

Reply:
General, `System.Exception`. The details of the exception are in the inner message from JS. What's the benefit of having specific type? We could make a `JSException` but what's the gain?